core/trino-main/src/main/java/io/trino/operator/output/UnnestingPositionsAppender.java

```diff
@@ -107,4 +136,95 @@ private IntArrayList mapPositions(IntArrayList positions, DictionaryBlock block)
         }
         return IntArrayList.wrap(positionArray);
     }
+
+    private static class DictionaryBlockBuilder
+    {
+        private static final int INSTANCE_SIZE = instanceSize(DictionaryBlockBuilder.class);
+        private final int initialEntryCount;
+        private Block dictionary;
+        private int[] dictionaryIds;
+        private int positionCount;
+        private boolean closed;
+
+        public DictionaryBlockBuilder()
+        {
+            this(1024);
+        }
+
+        public DictionaryBlockBuilder(int initialEntryCount)
+        {
+            this.initialEntryCount = initialEntryCount;
+            this.dictionaryIds = new int[0];
+        }
+
+        public boolean isEmpty()
+        {
+            return positionCount == 0;
+        }
+
+        public Block build()
+        {
+            return DictionaryBlock.create(positionCount, dictionary, dictionaryIds);
+        }
+
+        public long getRetainedSizeInBytes()
+        {
+            return INSTANCE_SIZE
+                    + (long) dictionaryIds.length * Integer.BYTES
+                    + (dictionary != null ? dictionary.getRetainedSizeInBytes() : 0);
+        }
+
+        public boolean canAppend(Block dictionary)
+        {
+            return !closed && (dictionary == this.dictionary || this.dictionary == null);
+        }
+
+        public void append(IntArrayList mappedPositions, Block dictionary)
+        {
+            checkArgument(canAppend(dictionary));
+            this.dictionary = dictionary;
+            ensureCapacity(positionCount + mappedPositions.size());
+            System.arraycopy(mappedPositions.elements(), 0, dictionaryIds, positionCount, mappedPositions.size());
+            positionCount += mappedPositions.size();
+        }
+
+        private void ensureCapacity(int capacity)
+        {
+            if (dictionaryIds.length >= capacity) {
+                return;
+            }
+
+            int newSize;
+            if (dictionaryIds.length > 0) {
+                newSize = calculateNewArraySize(dictionaryIds.length);
+            }
+            else {
+                newSize = initialEntryCount;
+            }
+            newSize = Math.max(newSize, capacity);
+
+            dictionaryIds = IntArrays.ensureCapacity(dictionaryIds, newSize, positionCount);
+        }
+
+        public void flushDictionary(PositionsAppender delegate)
+        {
+            if (closed) {
```

Comment:
Should we verify that `closed` is false here ?

Reply:
we then would have to do the `if !closed` on the outside every time, I think its better the way it is.